### ISSUE TITLE
[java] StringToString doesn't trigger on `Bar.class.getSimpleName().toString()`

### ISSUE THREAD

**Issue by ajeans** [pmd/pmd#3437]
**Affects PMD Version:** 

6.37

**Rule:** 

Please provide the rule name and a link to the rule documentation:

StringToString
<https://pmd.github.io/latest/pmd_rules_java_performance.html#stringtostring>

**Description:**

**Code Sample demonstrating the issue:**

```java
public class Foo {
  @Override
  public String toString() {
    return new StringJoiner(",", Bar.class.getSimpleName().toString() + "[", "]");
  }
}
```

**Expected outcome:**

PMD should report a violation at line 

`return new StringJoiner(",", Bar.class.getSimpleName().toString() + "[", "]")` 

but doesn't. This is a false-negative.

This seems to be documented in the Javadoc of the rule

> Note: due to an issue with type resolution, this implementation doesn't detect cases when toString() call is chained to a method returning String which is not declared in the class having the call or the call arguments are not of the exact same type as method parameters are, excluding the case when method name and number of it's parameters is enough to identify the method.

Is this limitation something that could be addressed with an additional Rule using XPath 2.0 ?

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by oowekyala** [comment]
> Is this limitation something that could be addressed with an additional Rule using XPath 2.0 ?

No, this is a limitation in our type resolution and XPath has less expressive access to that API as Java. So if a java rule cannot do it, an XPath rule couldn't either. Also I think adding new rules to fix limitations of existing ones should be avoided.

But thanks for the report, it's helpful as it will allow us to better track this documented limitation and fix it eventually.
